src/fm.py

```diff
@@ -1,36 +1,172 @@
 import argparse
 import sys
+import re
+
+def suffixArray(t):
+    """Given T return suffix array SA(T). 
+       We use Python's sorted function here 
+       for simplycity, but we can do better.
+
+    Args:
+        t (str): input string.
+    """  
+    satups = sorted([(t[i:], i) for i in range(len(t))])
+    
+    return list(map(lambda x: x[1], satups))
+
+def cigar_to_dict(cigar: str):
+    lst = [(int(i), op) for i, op in re.findall(r"(\d+)([^\d]+)", cigar)]
+    if(cigar!=''):
+        dict = {lst[i][1]: lst[i][0] for i in range(0, len(lst))}
+    return dict
+
+def dict_to_cigar(dict):
+    cigar = []
+    for key in dict:
+        cigar.append(dict[key])
+        cigar.append(key)
+    return ''.join(str(e) for e in cigar)
+
+def count_to_bucket(count: str):
+    '''
+    >>> cigar_to_edits("1$4i1m2p4s")
+    '0$1i5m6p8s'
+    '''
+    C = cigar_to_dict(count)
+    letter = list(C.keys())
+    bucket = {}
+    bucket[letter[0]] = 0
+    for i in range(1,len(letter)):
+        bucket[letter[i]] = bucket[letter[i-1]] + C[letter[i-1]]
+
+    return dict_to_cigar(bucket)
+    
+
+#BWT
+def bwt_C(x):
+    x += '$'
+    sa = suffixArray(x)
+    col = len(x)-1
+
+    bwt = ''.join([x[(i + col)%len(x)] for i in sa])
+    count = ''.join([x[i] for i in sa])
+    C = count_to_bucket(rle(count))         #  C is the bucket list
+    #O = calc_O(sa, C)
```

Comment:
I am not sure what you are doing with these (except that right now you are throwing them away), but `C` has nothing to do with a run length encoding. What you want in `C` is just a mapping from each letter to its offset in the sorted first column, so it is the accumulated sum from the radix sort exercise.